virtcontainers/qemu.go

```diff
@@ -825,6 +831,35 @@ func (q *qemu) waitSandbox(timeout int) error {
 		return err
 	}
 
+	if !q.config.UseVSock {
+		return nil
+	}
+
+	schemaInfo, err := q.qmpMonitorCh.qmp.ExecQueryQmpSchema(q.qmpMonitorCh.ctx)
+	if err != nil {
+		q.Logger().Error(err)
+		return err
+	}
+
+	wait := false
+	for _, schema := range schemaInfo {
+		if schema.MetaType == "event" && schema.Name == "VSOCK_RUNNING" {
+			wait = true
+		}
+	}
+
+	if !wait {
+		return nil
+	}
+
+	q.Logger().Info("waiting event of vsock running")
+	for {
+		ev := <-eventCh
```

Comment:
@ningbo9 Is there a possibility of a race here ie you may have missed the vosck "running" event while you are querying for the schema? If so, I would suggest listening on the channel in a separate go-routine that is started as early as possible.

Reply:
In the current PR, `query-qmp-schema` is almost the first QMP command (because `qmp_capabilities` must be executed before any other QMP commands are executed) and is called immediately after `LaunchQemu`, and `eventCh` receives the event until we read it.

I have no idea how to get this process earlier.